### PR TITLE
Add Week4

### PR DESCRIPTION
**Pull request by MarcRaeder**
Add Week4
- volume
- filter-less
- filter-more
-  recover

### CODE REVIEW

File: week3/plurality/plurality.c [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (candidates[i].votes == maximum_vote)
+        {
+            printf("%s\n", candidates[i].name);
+        }      
+    }
+    
+    return;
+}
```
Comment (matthiasdanne) [inline week3/plurality/plurality.c]: Add final newline.

File: week3/runoff/runoff.c [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (candidates[i].votes == min)
+        {
+            candidates[i].eliminated = true;
+        }
+    }
+    
+    return;
+}
```
Comment (matthiasdanne) [inline week3/runoff/runoff.c]: Add final newline.

File: week3/tideman/tideman.c [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        }
+
+        if (found_source)
+        {
+            printf("%s\n", candidates[i]);
+            return;
+        }
+    }
+    return;
+}
```
Comment (matthiasdanne) [inline week3/tideman/tideman.c]: Add final newline.

File: week4/filter-less/bmp.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+// The RGBTRIPLE structure describes a color consisting of relative intensities of
+// red, green, and blue. Adapted from http://msdn.microsoft.com/en-us/library/aa922590.aspx.
+
+typedef struct
+{
+    BYTE  rgbtBlue;
+    BYTE  rgbtGreen;
+    BYTE  rgbtRed;
+} __attribute__((__packed__))
+RGBTRIPLE;
```
Comment (matthiasdanne) [inline week4/filter-less/bmp.h]: Add final newline.

**Review by matthiasdanne — CHANGES_REQUESTED** [review]
- Isn't week3 already merged? 
- You should add a [.gitignore file](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files) and add `.DS_Store` there. These files are generated by Mac OS and you don't want to check them in.

File: week3/plurality/plurality.c [hunk 5]
```diff
@@ -0,0 +1,91 @@
+#include <cs50.h>
+#include <stdio.h>
+#include <string.h>
+
+#define MAX 9
```
Comment (DevBastiS) [inline week3/plurality/plurality.c]:
```suggestion
#define MAX_CANDIDATES 9
```

File: week3/plurality/plurality.c [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    print_winner();
+}
+
+bool vote(string name)
+{
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (strcmp(candidates[i].name, name) == 0)
```
Comment (DevBastiS) [inline week3/plurality/plurality.c]: Move to speaking variable.

File: week3/plurality/plurality.c (hunk 1 above)
Comment (DevBastiS) [inline week3/plurality/plurality.c]: Missing EOF.

File: week3/runoff/runoff.c [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        printf("Usage: runoff [candidate ...]\n");
+        return 1;
+    }
+
+    candidate_count = argc - 1;
+    if (candidate_count > MAX_CANDIDATES)
+    {
+        printf("Maximum number of candidates is %i\n", MAX_CANDIDATES);
+        return 2;
+    }
```
Comment (DevBastiS) [inline week3/runoff/runoff.c]:
```suggestion
    }
    
```

File: week3/runoff/runoff.c [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+            {
+                printf("Invalid vote.\n");
+                return 4;
+            }
+        }
+
+        printf("\n");
+    }
+
+    // Keep holding runoffs until winner exists
```
Comment (DevBastiS) [inline week3/runoff/runoff.c]: Remove comments if possible ;)

File: week3/runoff/runoff.c [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+        // Eliminate anyone with minimum number of votes
+        eliminate(min);
+
+        // Reset vote counts back to zero
+        for (int i = 0; i < candidate_count; i++)
+        {
+            candidates[i].votes = 0;
+        }
+    }
```
Comment (DevBastiS) [inline week3/runoff/runoff.c]:
```suggestion
    }
    
```

File: week3/runoff/runoff.c [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (strcmp(candidates[i].name, name) == 0)
+        {
+            preferences[voter][rank] = i;
+            return true;
+        }     
+    }
+    
+        return false;
```
Comment (DevBastiS) [inline week3/runoff/runoff.c]: Wrong indent

File: week3/runoff/runoff.c [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+    }
+    return 0;
+}
+
+// Record preference if vote is valid
+bool vote(int voter, int rank, string name)
+{
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (strcmp(candidates[i].name, name) == 0)
```
Comment (DevBastiS) [inline week3/runoff/runoff.c]: Move to speaking variable.

File: week3/runoff/runoff.c [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+int find_min(void)
+{
+    int min = voter_count;
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (candidates[i].votes < min && candidates[i].eliminated == false) 
+        {
+            min = candidates[i].votes;
+        }
+
```
Comment (DevBastiS) [inline week3/runoff/runoff.c]:
```suggestion
```

File: week3/tideman/tideman.c [hunk 13]
```diff
@@ -0,0 +1,226 @@
+#include <cs50.h>
+#include <stdio.h>
+#include <string.h>
+
+#define MAX 9
```
Comment (DevBastiS) [inline week3/tideman/tideman.c]:
```suggestion
#define MAX_CANDIDATES 9
```

File: week3/tideman/tideman.c [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+        if (!found)
+        {
+            winner = -1;
+        }
+    }
+
+    if (winner == loser)
+    {
+        return true;
+    }
```
Comment (DevBastiS) [inline week3/tideman/tideman.c]:
```suggestion
    }
    
```

File: week3/tideman/tideman.c [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+        printf("Usage: tideman [candidate ...]\n");
+        return 1;
+    }
+
+    candidate_count = argc - 1;
+    if (candidate_count > MAX)
+    {
+        printf("Maximum number of candidates is %i\n", MAX);
+        return 2;
+    }
```
Comment (DevBastiS) [inline week3/tideman/tideman.c]:
```suggestion
    }
    
```

File: week3/tideman/tideman.c [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+
+void lock_pairs(void)
+{
+    for (int i = 0; i < pair_count; i++)
+    {
+        if (!hasCycle(pairs[i].winner, pairs[i].loser))
+        {
+            locked[pairs[i].winner][pairs[i].loser] = true;
+        }
+    }
```
Comment (DevBastiS) [inline week3/tideman/tideman.c]:
```suggestion
    }
    
```

**Review by DevBastiS — CHANGES_REQUESTED** [review]

File: week4/filter-less/helpers.c [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    RGBTRIPLE copy[height][width];
+
+    for (int i = 0; i < height; i++)
+    {
+        for (int j = 0; j < width; j++)
+        {
+            copy[i][j] = image[i][j];
+        }
+
+        for (int i = 0; i < height; i++)
```
Comment (matthiasdanne) [inline week4/filter-less/helpers.c]: - You use the same variable inside a nested scope. This is called variable shadowing and should be avoided.
- Your `blur` function is rather long and has many for-loops. You could move some of your code to helper functions with meaningful names, to increase the understandability of your code.

**Review by matthiasdanne — CHANGES_REQUESTED** [review]

**Review by matthiasdanne — CHANGES_REQUESTED** [review]
Overall these are good solutions. ✅

- You could split your logic into smaller functions, to increase readability.
- Add a hint that the sobel operator was used to calculate edges. (maybe add a function `applySobelOperator`)

File: week4/filter-more/helpers.c [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+            image[i][j].rgbtGreen = temp[i][j].rgbtGreen;
+            image[i][j].rgbtRed = temp[i][j].rgbtRed;
+        }
+    }
+
+    return;
+}
+
+
+
```
Comment (matthiasdanne) [inline week4/filter-more/helpers.c]:
```suggestion
```

File: week4/filter-more/helpers.c [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+    for (int i = 0; i < height; i++)
+    {
+        for (int j = 0; j < width; j++)
+        {
+            image[i][j].rgbtBlue = temp[i][j].rgbtBlue;
+            image[i][j].rgbtGreen = temp[i][j].rgbtGreen;
+            image[i][j].rgbtRed = temp[i][j].rgbtRed;
+        }
+    }
```
Comment (matthiasdanne) [inline week4/filter-more/helpers.c:171]: Couldn't you just call `copy_gets_image(height, width, image, temp)` here?

**Review by matthiasdanne — CHANGES_REQUESTED** [review]

File: week4/filter-less/helpers.c [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+        {
+            for (int j = 0; j < width; j++)
+            {
+                image[i][j].rgbtRed = copy[i][j].rgbtRed;
+                image[i][j].rgbtGreen = copy[i][j].rgbtGreen;
+                image[i][j].rgbtBlue = copy[i][j].rgbtBlue;
+            }
+        }  
+
+        return;
```
Comment (matthiasdanne) [inline week4/filter-less/helpers.c]:
```suggestion
    return;
```

**Review by matthiasdanne — APPROVED** [review]

**Review by stefanluth — CHANGES_REQUESTED** [review]
See Louisa's PR :D

File: week4/filter-less/helpers.c [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+
+                    copy[i][l].rgbtRed = round(totalRed / counter);
+                    copy[i][l].rgbtGreen = round(totalGreen / counter);
+                    copy[i][l].rgbtBlue = round(totalBlue / counter);
+                }
+            }
+        }
+
+        
+    }
```
Comment (stefanluth) [inline week4/filter-less/helpers.c]:
```suggestion
        }
    }

```

File: week4/recover/recover.c [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+
+    char *filename = malloc(8 * sizeof(char));
+    int count_image = 0;
+    unsigned char buffer[512];
+
+    FILE *output_file = NULL;
+
+    while (fread(buffer, sizeof(char), 512, input_file))
+    {
+        bool checkBytesIndicateStartOfJPEG = (buffer[0] == 0xff && buffer[1] == 0xd8 && buffer[2] == 0xff && (buffer[3] & 0xff) == 0xe0);
```
Comment (stefanluth) [inline week4/recover/recover.c]: Keep variable names consistent. (snake_case)

**Review by stefanluth — CHANGES_REQUESTED** [review]
See [Louisa's PR](https://github.com/LouisaStrunk02/CS50Kurs/pull/9).

File: week4/filter-less/helpers.c [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+
+void grayscale(int height, int width, RGBTRIPLE image[height][width])
+{
+    for (int i = 0; i < height; i++)
+    {
+        for (int j = 0; j < width; j++)
+        {
+            float Red = image[i][j].rgbtRed;
+            float Green = image[i][j].rgbtGreen;
+            float Blue = image[i][j].rgbtBlue;
```
Comment (stefanluth) [inline week4/filter-less/helpers.c]:
```suggestion
            float red = image[i][j].rgbtRed;
            float green = image[i][j].rgbtGreen;
            float blue = image[i][j].rgbtBlue;
```

File: week4/filter-less/helpers.c [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+            for (int j = 0; j < width; j++)
+            {
+                image[i][j].rgbtRed = copy[i][j].rgbtRed;
+                image[i][j].rgbtGreen = copy[i][j].rgbtGreen;
+                image[i][j].rgbtBlue = copy[i][j].rgbtBlue;
+            }
+        }  
+
+        return;
+}
```
Comment (stefanluth) [inline week4/filter-less/helpers.c]: Indentation is off.

File: week4/filter-more/helpers.c [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+
+void grayscale(int height, int width, RGBTRIPLE image[height][width])
+{
+    for (int i = 0; i < height; i++)
+    {
+        for (int j = 0; j < width; j++)
+        {
+            float Red = image[i][j].rgbtRed;
+            float Green = image[i][j].rgbtGreen;
+            float Blue = image[i][j].rgbtBlue;
```
Comment (stefanluth) [inline week4/filter-more/helpers.c]: Lowercase variables

File: week4/filter-more/helpers.c [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+        {
+            for (int j = 0; j < width; j++)
+            {
+                image[i][j].rgbtRed = copy[i][j].rgbtRed;
+                image[i][j].rgbtGreen = copy[i][j].rgbtGreen;
+                image[i][j].rgbtBlue = copy[i][j].rgbtBlue;
+            }
+        }  
+
+        return;
```
Comment (stefanluth) [inline week4/filter-more/helpers.c]: Indentation

File: week4/filter-more/helpers.c [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+{
+    for (int i = 0; i < height; i++)
+    {
+        for (int j = 0; j < width; j++)
+        {
+            image[i][j].rgbtBlue = temp[i][j].rgbtBlue;
+            image[i][j].rgbtGreen = temp[i][j].rgbtGreen;
+            image[i][j].rgbtRed = temp[i][j].rgbtRed;
+        }
+    }
```
Comment (stefanluth) [inline week4/filter-more/helpers.c:171]: Missing return

File: week4/volume/volume.c [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+
+    int16_t buffer;
+    while (fread(&buffer, sizeof(int16_t), 1, input))
+    {
+        buffer *= factor;
+        fwrite(&buffer, sizeof(int16_t), 1, output);
+    }
+    
+    fclose(input);
+    fclose(output);
```
Comment (stefanluth) [inline week4/volume/volume.c:43]: Missing return

**Review by stefanluth — CHANGES_REQUESTED** [review]

File: week4/filter-more/helpers.c [hunk 29]
```diff
@@ -0,0 +1,175 @@
...
+    {
+        for (int j = 0; j < width; j++)
+        {
+            image[i][j].rgbtRed = copy[i][j].rgbtRed;
+            image[i][j].rgbtGreen = copy[i][j].rgbtGreen;
+            image[i][j].rgbtBlue = copy[i][j].rgbtBlue;
+        }
+    }  
+
+        return;
```
Comment (stefanluth) [inline week4/filter-more/helpers.c]: Indentation

**Review by stefanluth — APPROVED** [review]

**Pull request merged**